group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/CurrentAssignmentBuilder.java

```diff
@@ -0,0 +1,415 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.consumer;
+
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.message.ConsumerGroupHeartbeatRequestData;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiFunction;
+
+/**
+ * The CurrentAssignmentBuilder class encapsulates the reconciliation engine of the
+ * consumer group protocol. Given the current state of a member and a desired or target
+ * assignment state, the state machine takes the necessary steps to converge them.
+ *
+ * The member state has the following properties:
+ * - Current Epoch  - The current epoch of the member.
+ * - Next Epoch     - The desired epoch of the member. It corresponds to the epoch of
+ *                    the target/desired assignment. The member transitions to this epoch
+ *                    when it has revoked the partitions that it does not own or if it
+ *                    does not have to revoke any.
+ * - Previous Epoch - The previous epoch of the member when the state was updated.
+ * - Assigned Set   - The set of partitions currently assigned to the member. This represents what
+ *                    the member should have.
+ * - Revoking Set   - The set of partitions that the member should revoke before it can transition
+ *                    to the next state.
+ * - Assigning Set  - The set of partitions that the member will eventually receive. The partitions
+ *                    in this set are still owned by other members in the group.
+ *
+ * The state machine has four states:
+ * - NEW_TARGET_ASSIGNMENT - This is the initial state of the state machine. The state machine starts
+ *                           here when the next epoch does not match the target epoch. It means that
+ *                           a new target assignment has been installed so the reconciliation process
+ *                           must restart. In this state, the Assigned, Revoking and Assigning sets
+ *                           are computed. If Revoking is not empty, the state machine transitions
+ *                           to REVOKE; if Assigning is not empty, it transitions to ASSIGNING;
+ *                           otherwise it transitions to STABLE.
+ * - REVOKE                - This state means that the member must revoke partitions before it can
+ *                           transition to the next epoch and thus start receiving new partitions.
+ *                           The member transitions to the next state only when it has acknowledged
+ *                           the revocation.
+ * - ASSIGNING             - This state means that the member waits on partitions which are still
```

Comment:
same here, assigning gives the impression that the member is assigning something?

Reply:
The coordinator is actually assigning partitions to the member in this state.